File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-msg-5-4/PortfolioReferenceBase.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class PortfolioReferenceBase : ISerialized
    {
        public PortfolioReferenceBase(XmlNode xmlNode)
        {
            XmlNode portfolioNameNode = xmlNode.SelectSingleNode("portfolioName");
            
            if (portfolioNameNode != null)
            {
                if (portfolioNameNode.Attributes["href"] != null || portfolioNameNode.Attributes["id"] != null) 
                {
                    if (portfolioNameNode.Attributes["id"] != null) 
                    {
                        portfolioNameIDRef_ = portfolioNameNode.Attributes["id"].Value;
                        PortfolioName ob = new PortfolioName(portfolioNameNode);
                        IDManager.SetID(portfolioNameIDRef_, ob);
                    }
                    else if (portfolioNameNode.Attributes["href"] != null)
                    {
                        portfolioNameIDRef_ = portfolioNameNode.Attributes["href"].Value;
                    }
                    else
                    {
                        portfolioName_ = new PortfolioName(portfolioNameNode);
                    }
                }
                else
                {
                    portfolioName_ = new PortfolioName(portfolioNameNode);
                }
            }
            
        
        }
        
    
        #region PortfolioName_
        private PortfolioName portfolioName_;
        public PortfolioName PortfolioName_
        {
            get
            {
                if (this.portfolioName_ != null)
                {
                    return this.portfolioName_; 
                }
                else if (this.portfolioNameIDRef_ != null)
                {
                    portfolioName_ = IDManager.getID(portfolioNameIDRef_) as PortfolioName;
                    return this.portfolioName_; 
                }
                else
                {
                      return this.portfolioName_; 
                }
            }
            set
            {
                if (this.portfolioName_ != value)
                {
                    this.portfolioName_ = value;
                }
            }
        }
        #endregion
        
        public string portfolioNameIDRef_ { get; set; }
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-msg-5-4/PortfolioReferenceBase.cpp
// PortfolioReferenceBase.cpp 
#include "PortfolioReferenceBase.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

PortfolioReferenceBase::PortfolioReferenceBase(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //portfolioNameNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* portfolioNameNode = xmlNode->FirstChildElement("portfolioName");

   if(portfolioNameNode){portfolioNameIsNull_ = false;}
   else{portfolioNameIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- portfolioNameNode , address : " << portfolioNameNode << std::endl;
   #endif
   if(portfolioNameNode)
   {
      if(portfolioNameNode->Attribute("href") || portfolioNameNode->Attribute("id"))
      {
          if(portfolioNameNode->Attribute("id"))
          {
             portfolioNameIDRef_ = portfolioNameNode->Attribute("id");
             portfolioName_ = boost::shared_ptr<PortfolioName>(new PortfolioName(portfolioNameNode));
             portfolioName_->setID(portfolioNameIDRef_);
             IDManager::instance().setID(portfolioNameIDRef_,portfolioName_);
          }
          else if(portfolioNameNode->Attribute("href")) { portfolioNameIDRef_ = portfolioNameNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { portfolioName_ = boost::shared_ptr<PortfolioName>(new PortfolioName(portfolioNameNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<PortfolioName> PortfolioReferenceBase::getPortfolioName()
{
   if(!this->portfolioNameIsNull_){
        if(portfolioNameIDRef_ != ""){
             return boost::shared_static_cast<PortfolioName>(IDManager::instance().getID(portfolioNameIDRef_));
        }else{
             return this->portfolioName_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<PortfolioName>();
   }
}
}

